pydantic/_internal/_dataclasses.py

```diff
@@ -123,19 +124,23 @@ def complete_dataclass(
         typevars_map,
     )
 
-    # dataclass.__init__ must be defined here so its `__qualname__` can be changed since functions can't be copied.
+    # Get a temporary signature before we change the __init__
+    fields = cls.__pydantic_fields__  # type: ignore
+    # This needs to be called before we change the __init__
+    sig = generate_dataclass_signature(cls, fields, config_wrapper)
```

Comment:
```suggestion
    sig = generate_dataclass_signature(cls, cls.__pydantic_fields__, config_wrapper)
```
I think we can simplify this as such, your first comment is clear enough :)